src/orange/orangeinp/RevolvedPolygon.cc

```diff
@@ -0,0 +1,273 @@
+/*!
+ * Construct a volume from this shape.
+ */
+NodeId RevolvedPolygon::build(VolumeBuilder& vb) const
+{
+    // Use the volume builder's tolerance to remove any colinear points
+    auto filtered_polygon
+        = detail::filter_collinear_points(polygon_, vb.tol().abs);
+
+    // After removing collinear points, at least 3 points must remain
+    CELER_VALIDATE(filtered_polygon.size() >= 3,
+                   << "polygon must consist of at least 3 points");
+
+    // Start the recursion process
+    SubIndex ri;
+    return this->make_levels(vb, filtered_polygon, ri);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write the shape to JSON.
+ */
+void RevolvedPolygon::output(JsonPimpl* j) const
+{
+    to_json_pimpl(j, *this);
+}
+//-------------------------------------------------------------------------//
+// HELPER FUNCTION DEFINITIONS
+//-------------------------------------------------------------------------//
+/*!
+ * Recursively construct convex regions, subtracting out concavities.
+ */
+NodeId RevolvedPolygon::make_levels(detail::VolumeBuilder& vb,
+                                    VecReal2 const& polygon,
+                                    RevolvedPolygon::SubIndex si) const
+{
+    // Find the convex hull
+    detail::ConvexHullFinder<real_type> hull_finder(polygon, vb.tol());
+    auto convex_hull = hull_finder.make_convex_hull();
+    auto concave_regions = hull_finder.calc_concave_regions();
+
+    // Build the convex region
+    auto filtered_convex_hull
+        = detail::filter_collinear_points(convex_hull, vb.tol().abs);
+    NodeId result = this->make_region(vb, filtered_convex_hull, si);
+
+    // Return early if there are no concave regions to process
+    if (concave_regions.empty())
+    {
+        return result;
+    }
+
+    // Create a vector of all concave regions, via recursion
+    std::vector<NodeId> concave_nodes(concave_regions.size());
+    for (auto i : range(static_cast<size_type>(concave_nodes.size())))
+    {
+        concave_nodes[i] = this->make_levels(
+            vb, concave_regions[i], SubIndex{si.level + 1, i});
+    }
+
+    auto level_ext = this->make_level_ext(si);
+
+    // Create a union of all concave regions
+    NodeId concave_union
+        = vb.insert_region(Label{label_, level_ext + ".cu"},
+                           Joined{op_or, std::move(concave_nodes)});
+
+    // Create a negation of this union
+    auto sub_node = vb.insert_region(Label{label_, level_ext + ".ncu"},
+                                     Negated{concave_union});
+
+    // Subtract concave regions from the convex hull
+    return vb.insert_region(Label{label_, level_ext + ".d"},
+                            Joined{op_and, {result, sub_node}});
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Revolved a convex polygon around the \em z axis.
+ *
+ * The polygon be strictly counterclockwise, i.e., it must not contain any
```

Comment:
```suggestion
 * The polygon must be strictly counterclockwise, i.e., it must not contain any
```

Unless ye be a pirate